compiler/rustc_mir_transform/src/generator.rs

```diff
@@ -926,20 +957,26 @@ fn compute_layout<'tcx>(
         debug!(?decl);
 
         let ignore_for_traits = if tcx.sess.opts.unstable_opts.drop_tracking_mir {
-            // Do not `assert_crate_local` here, as post-borrowck cleanup may have already cleared
-            // the information. This is alright, since `ignore_for_traits` is only relevant when
-            // this code runs on pre-cleanup MIR, and `ignore_for_traits = false` is the safer
-            // default.
-            match decl.local_info {
-                // Do not include raw pointers created from accessing `static` items, as those could
-                // well be re-created by another access to the same static.
-                ClearCrossCrate::Set(box LocalInfo::StaticRef { is_thread_local, .. }) => {
-                    !is_thread_local
+            if !init_locals.contains(local) {
+                // If only the storage is required to be live, but local is not initialized, then we can
+                // ignore such type for auto trait purposes.
+                true
```

Comment:
Why do we require storage to be live if the locals is not initialized? Should the general algorithm be changed to avoid requiring storage? That'd kill 2 birds with one stone: the spurious auto trait limitation and the generator size.

As a general rule, I'd rather keep the "storage" and the "traits" algorithm as close to each other as possible: any deviation between the two is a potential source of unsoundness. This exact change is probably not, but the general rule should stand.

Reply:
That's currently an open question rust-lang/unsafe-code-guidelines#188 and is much harder to solve.